### PR TITLE
ImageProcessing: expose some functions and allows in-memory processing

### PR DESCRIPTION
**Pull request by CyaNinix**
Currently, most of image processing is done inside the ImageProcessingAtom gem, and only a few functions are exposed.

It's interesting to expose some other functions to be able to do some image processing in memory, for example for a texture array  gem which only write one file for multiple images.

### CODE REVIEW

File: Gems/Atom/Asset/ImageProcessingAtom/Code/Include/Atom/ImageProcessing/ImageProcessingBus.h [hunk 1]
```diff
@@ -64,11 +65,30 @@ namespace ImageProcessingAtom
             const AZ::Data::AssetId& sourceAssetId,
             const AZStd::string& sourceAssetName) = 0;
 
+        virtual IImageObjectPtr ConvertImageObjectInMemory(
+            IImageObjectPtr imageObject,
+            const AZStd::string& presetName,
+            const AZStd::string& platformName,
+            const AZ::Data::AssetId& sourceAssetId,
+            const AZStd::string& sourceAssetName) = 0;
```
Comment (gadams3) [inline Gems/Atom/Asset/ImageProcessingAtom/Code/Include/Atom/ImageProcessing/ImageProcessingBus.h:73]: Please add comments to this function

File: Gems/Atom/Asset/ImageProcessingAtom/Code/Include/Atom/ImageProcessing/ImageProcessingBus.h [hunk 2]
```diff
@@ -64,11 +65,30 @@ namespace ImageProcessingAtom
...
+
+        virtual AZStd::vector<AZStd::string> GetFileMasksForPreset(const PresetName& presetName) = 0;
+
+        virtual AZStd::vector<PresetName> GetPresetsForFileMask(const FileMask& fileMask) = 0;
+
+        virtual PresetName GetDefaultPreset() = 0;
+
+        virtual PresetName GetDefaultAlphaPreset() = 0;
+
+        virtual bool IsValidPreset(PresetName presetName) = 0;
```
Comment (gadams3) [inline Gems/Atom/Asset/ImageProcessingAtom/Code/Include/Atom/ImageProcessing/ImageProcessingBus.h:91]: Please comment these functions, inputs and outputs

File: Gems/Atom/Asset/ImageProcessingAtom/Code/Source/BuilderSettings/BuilderSettingManager.cpp [hunk 3]
```diff
@@ -605,6 +605,30 @@ namespace ImageProcessingAtom
...
+
+    AZStd::vector<PresetName> BuilderSettingManager::GetPresetsForFileMask(const FileMask& fileMask) const
+    {
+        AZStd::vector<PresetName> presets;
+
+        AZStd::lock_guard<AZStd::recursive_mutex> lock(m_presetMapLock);
+        if (auto it = m_presetFilterMap.find(fileMask); it != m_presetFilterMap.end())
+        {
+            for (const PresetName& presetName : it->second)
+                presets.push_back(presetName);
```
Comment (gadams3) [inline Gems/Atom/Asset/ImageProcessingAtom/Code/Source/BuilderSettings/BuilderSettingManager.cpp:626]: Please add braces to all loops and conditionals

File: Gems/Atom/Asset/ImageProcessingAtom/Code/Source/BuilderSettings/BuilderSettingManager.h [hunk 4]
```diff
@@ -102,6 +102,14 @@ namespace ImageProcessingAtom
...
+
+        PresetName GetDefaultPreset() const;
+
+        PresetName GetDefaultAlphaPreset() const;
+
+        AZStd::vector<PresetName> GetPresetsForFileMask(const FileMask& fileMask) const;
+
         bool IsValidPreset(PresetName presetName) const;
 
         bool DoesSupportPlatform(AZStd::string_view platformId);
```
Comment (gadams3) [inline Gems/Atom/Asset/ImageProcessingAtom/Code/Source/BuilderSettings/BuilderSettingManager.h:115]: Please add 'override' to all of these overridden virtual function definitions.

File: Gems/Atom/Asset/ImageProcessingAtom/Code/Source/ImageBuilderComponent.cpp [hunk 5]
```diff
@@ -196,6 +196,46 @@ namespace ImageProcessingAtom
...
+        desc->m_sourceAssetId = sourceAssetId;
+        
+        // Create an image convert process
+        ImageConvertProcess process(AZStd::move(desc));
+        process.ProcessAll();
+        bool result = process.IsSucceed();
+        if (result)
+            return process.GetOutputImage();
+        else
+            return nullptr;
```
Comment (gadams3) [inline Gems/Atom/Asset/ImageProcessingAtom/Code/Source/ImageBuilderComponent.cpp:236]: Please add braces

**Review by gadams3 — CHANGES_REQUESTED** [review]

File: Gems/Atom/Asset/ImageProcessingAtom/Code/Source/BuilderSettings/BuilderSettingManager.h (hunk 4 above)
Comment (CyaNinix) [reply Gems/Atom/Asset/ImageProcessingAtom/Code/Source/BuilderSettings/BuilderSettingManager.h:115]: Those are not virtual functions.

File: Gems/Atom/Asset/ImageProcessingAtom/Code/Source/ImageBuilderComponent.cpp [hunk 6]
```diff
@@ -196,6 +196,50 @@ namespace ImageProcessingAtom
...
+        desc->m_platform = platformName;
+        desc->m_filePath = presetFilePath;
+        desc->m_isStreaming = BuilderSettingManager::Instance()->GetBuilderSetting(platformName)->m_enableStreaming;
+        desc->m_imageName = sourceAssetName;
+        desc->m_shouldSaveFile = false;
+        desc->m_sourceAssetId = sourceAssetId;
+        
+        // Create an image convert process
+        ImageConvertProcess process(AZStd::move(desc));
+        process.ProcessAll();
```
Comment (santorac) [inline Gems/Atom/Asset/ImageProcessingAtom/Code/Source/ImageBuilderComponent.cpp:231]: Can you factor this out to share code with ConvertImageObject?

**Review by santorac — APPROVED** [review]
These changes look good to me. There is one place where I would like to see the code factored out, but it's not significant enough to block the PR so you have my approval either way.

File: Gems/Atom/Asset/ImageProcessingAtom/Code/Source/ImageBuilderComponent.cpp (hunk 6 above)
Comment (CyaNinix) [reply Gems/Atom/Asset/ImageProcessingAtom/Code/Source/ImageBuilderComponent.cpp:231]: I can, but this requires me to either forward declare ImageConvertProcess in ImageBuilderComponent.h (to declare a private function using it), or make an anonymous function inside ImageBuilderComponent.cpp.

**Review by cgalvan — APPROVED** [review]

**Pull request merged**